### PR TITLE
Add `-transition=in` compiler flag to help users transition away from using `in`

### PR DESCRIPTION
**Pull request by JinShil**
A "gentler" alternative to https://github.com/dlang/dmd/pull/10179

`in` as a parameter storage class is defined as `scope const`.  However `in` has not yet
been properly implemented so its current implementation is equivalent to `const`.  Properly
implementing `in` now will likely break code, so it is recommended to avoid using `in`, and
explicitly use `const` or `scope const` instead, until `in` is properly implemented.

The use of `in` as a parameter storage class is already discouraged in the documentation.  See https://dlang.org/spec/function.html#parameters

### CODE REVIEW

File: src/dmd/parse.d [hunk 1]
```diff
@@ -2901,6 +2901,13 @@ final class Parser(AST) : Lexer
                     }
                 case TOK.in_:
                     stc = STC.in_;
+                    if (global.params.vin)
+                    {
+                        deprecation("`in` is defined as `scope const`.  However, `in` has not yet been properly " ~
+                            "implemented, so its current implementation is equivalent to `const`. It is recommended " ~
+                            "to avoid using `in`, and explicitly use `scope const` or `const` instead, until `in` " ~
+                            "is properly implemented.");
```
Comment (wilzbach) [inline src/dmd/parse.d]: I think this is a bit too long for a deprecation message that might be displayed hundreds of times.

Comment (JinShil) [reply src/dmd/parse.d]: How about now?

Comment (rainers) [reply src/dmd/parse.d]: An alternative could be to emit a longer version only once, and a shorter message for other occurrences.

**Comment by rainers** [comment]
Should the message also be emitted for disabled code? If not, the check needs to be done during semantic analysis.

**Comment by 12345swordy** [comment]
I still don't understand why we can't simply redefine "in" as "const scope" when -dip1000 flag has been applied. Deprecating "in" seems to me a bad decision giving that "inout" and "out" still exist.

**Comment by JinShil** [comment]
Please understand that we're not deprecating `in`.  All we're doing is telling users not to use it yet, because if and when it is implemented it could break their code.  Also, there are a few other ideas floating around to make in more than just `const scope` (i.e. `const scope ref`).  It's just best if users don't use `in` until a decision can be made about it.

Right now we're sabotaging users by not informing them to stay away from `in`.  And because users are using it, it limits design possibilities due to the fact that it may disrupt users.  If we can just get the word out to not use `in` we can define it as we wish, and users can start utilizing it when its ready.

**Comment by JinShil** [comment]
> Should the message also be emitted for disabled code?

I argue it should, because the goal here is to inform the community that the future of `in` is uncertain, and to stop using it altogether until it's ready.

File: src/dmd/parse.d (hunk 1 above)
Comment (JinShil) [reply src/dmd/parse.d]: That sounds like a good idea.  I think I'll try to implement that tomorrow.

Comment (rainers) [reply src/dmd/parse.d]: Probably simplest with a supplemental deprecation message.

**Comment by rainers** [comment]
> I argue it should, because the goal here is to inform the community that the future of `in` is uncertain, and to stop using it altogether until it's ready.

I was thinking about projects that try to support a range of compiler versions using conditional compilation for constructs that work differently across different versions. `in` can probably still be replaced by `const` for older versions without semantic change, though.

**Review by Geod24 — CHANGES_REQUESTED** [review]
My objection still stands: We should leave it as is before we have a proper plan for it. Using `transition` means we are going to change something, in a way that can't be easily deprecated. Here we don't have a plan of timeline for it.

I was not a fan of changing every instance in Phobos & druntime, because it felt like just noise, but it's still things under our control so it matters less than affecting *all the D code out there*.

File: src/dmd/parse.d (hunk 1 above)
Comment (JinShil) [reply src/dmd/parse.d]: Done.  Thank you.

**Comment by JinShil** [comment]
- Changed the message from deprecation to warning
- Move the logic to semantic so only the module being compiled emits messages.
- A more verbose supplemental message is displayed once for the first occurrence of `in` and omitted for any additional occurrences.

I still think this is the right thing to avoid sabotaging users and to allow designers the liberty to design and implement `in` right without having to worry too much about disruption.

**Comment by JinShil** [comment]
No longer WIP, and ready to go.

**Comment by atilaneves** [comment]
> My objection still stands: We should leave it as is before we have a proper plan for it. Using `transition` means we are going to change something, in a way that can't be easily deprecated. Here we don't have a plan of timeline for it.
> 
> I was not a fan of changing every instance in Phobos & druntime, because it felt like just noise, but it's still things under our control so it matters less than affecting _all the D code out there_.

There's a plan for it. The plan is to make `in` mean `const scope` as intended when dip1000 is enabled. I was working on this last week, I just didn't manage to figure out exactly how to change the compiler to do it in the time I had for the task.

**Comment by JinShil** [comment]
> The plan is to make in mean const scope as intended when dip1000 is enabled.

Sounds good to me.

**Comment by Geod24** [comment]
> There's a plan for it.

Then there's no need to discourage its usage, instead just provide a transitional switch to make it means what it was supposed to mean. Or pack it with `dip1000`, whatever you think is best.
And what happens to all the places where it was removed ? Should we revert those ?

**Comment by atilaneves** [comment]
> > There's a plan for it.
> 
> Then there's no need to discourage its usage,

Precisely.

> instead just provide a transitional switch to make it means what it was supposed to mean. Or pack it with `dip1000`, whatever you think is best.

The plan is to pack it in with dip1000.

> And what happens to all the places where it was removed ? Should we revert those ?

I didn't think they should have been removed at the time, but I had to talk to Walter first to know if he'd agree with my plan.

**Comment by JinShil** [comment]
@atilaneves Once `in` becomes `scope const`, some things will not compile.  See https://github.com/dlang/druntime/pull/2685#discussion_r304370819 and https://issues.dlang.org/show_bug.cgi?id=20087 for a couple of examples.

**Comment by atilaneves** [comment]
@JinShil Yes, but only with DIP1000, which is optional and creates a time period for not breaking. Do you have any suggestions on how to handle those two breakages?

**Comment by JinShil** [comment]
>  Do you have any suggestions on how to handle those two breakages?

I don't know what to do about https://github.com/dlang/druntime/pull/2685#discussion_r304370819 yet, but perhaps you could get @WalterBright on the horn and ask him if https://github.com/dlang/dmd/pull/10223 could be a solution to [Issue 20087](https://issues.dlang.org/show_bug.cgi?id=20087).

I also might be able to do the implementation of `in` for you, but I might need a couple of weeks.  I was always willing to do the work, I just needed a decision.

**Comment by JinShil** [comment]
@atilaneves Also, [this](https://github.com/dlang/druntime/pull/2782#discussion_r319742948) is another situation that doesn't build.  If you could tell me how the compiler *should* behave, I might be able to fix it.

**Comment by ZombineDev** [comment]
@atilaneves can we also have `void foo(in T x)` mean `void foo(auto ref scope const(T) x)` with `-preview=dip1000` and `-preview=rvaluerefparam`? That way `in` parameters will be the logic counterpart of `out` parameters. The non-template `auto ref` meaning pass small types by value and large types by reference, unlike the [existing `auto ref`](https://run.dlang.io/gist/run-dlang/e79138f097d2a7955f729c3a2c7fdbe9?compiler=dmd) which takes lvalues always by ref and rvalues by value.

**Comment by atilaneves** [comment]
@ZombineDev I'd have to think about that. A lot. I assume by `in T x` you mean in a template function? I clicked on the two links but am unsure of what exactly is being proposed.

**Comment by Geod24** [comment]
```
struct Huge { /* ... */ }
void foo(in Huge val); // This is equivalent to `foo(const scope ref Huge val)`
void foo (in ushort);    // This is equivalent to `foo(const scope ushort)`
```

Basically, if you're passing an `in` parameter, it's an input to the function which will only be valid for the duration of the call and we should guarantee that copy don't happen.

**Comment by ZombineDev** [comment]
> I'd have to think about that. A lot. I assume by in T x you mean in a template function?

Nope, I meant both template and non-template functions. Currently `auto ref` is a strange thing. You can use it only with template functions - `void foo(T)(auto ref T x)`
What's even stranger is that you can't instantiate such templates explicitly, e.g. `auto fptr = &foo!T` doesn't work. That's because the only way to instantiate a function template with `auto ref` parameters is by actually calling the function with some arguments. That's because the compiler needs to instantiate either the `ref` or non-`ref` version depending on the run-time arguments provided at the call site.
For the same reason, you can't have virtual methods that take `auto ref` parameters, nor you can define function pointer or delegate types ("callbacks") that accept arguments by `auto ref`.

What I propose is (I hope) much simpler.
The `in` parameter attribute in a non-template function `void bar(in T x)` should be replaced during semantic analysis by either `scope const(T)`, or `scope ref const(T)`, depending on what is the most efficient way of passing the argument of type `T`. Small POD types should be copied and large types should be passed by reference. It should accept lvalues and rvalues alike (that's the whole point of the `-preview=rvaluerefparam` flag).

You would be able to define both template and non-template functions, virtual methods, function pointer and delegate types, etc. just like you can today.

In short, I think that `auto ref` is a dirty and inefficient hack, and with my proposal I hope we would get a much better alternative.

**Comment by Geod24** [comment]
Interestingly I had the exact same discussion with @JinShil a few weeks ago. Should there be a DIP for that ?

**Comment by atilaneves** [comment]
> ```d
> struct Huge { /* ... */ }
> void foo(in Huge val); // This is equivalent to `foo(const scope ref Huge val)`
> void foo (in ushort);    // This is equivalent to `foo(const scope ushort)`
> ```
> 
> Basically, if you're passing an `in` parameter, it's an input to the function which will only be valid for the duration of the call and we should guarantee that copy don't happen.

I think that a copy happening or not is orthogonal to it being an `in` parameter. One could achieve the same thing as now by labeling that parameter as `in ref`, assuming `in` gets implemented as intended.

**Comment by atilaneves** [comment]
@Geod24 Yes, I think a DIP would be needed.
